### ISSUE TITLE
"mathjax" option needed to turn off markdown when text surround by "$" "$"

### ISSUE THREAD

**Issue by wm2015email** [showdownjs/showdown#693]
Any chance you could add an option to turnoff markdown conversion between the tags:

$ ... $

and 

$$ ... $$

This is required for markdown using mathjax…  such as on math.stackexchange.com... see markdown used on this website...

I just noticed that when I convert markdown with embedded mathjax… some of the equations get messed up by markdown tags being converted inside of math equation fence characters of "$" and "$$"

would be nice to have this as an option.  example:

    converter = new showdown.Converter({underline: true, mathjax:true});

one example was this:

This is the markdown of the mathjax before conversion...
$$
\delta(n) = \begin{cases}
1 & n = 0 \\
0 & otherwise 
\end{cases}
$$
	
after conversion it converts "&" into &amp; ... mathjax doesn't like this... :

<p>$$
\delta(n) = \begin{cases}
1 &amp; n = 0 \
0 &amp; otherwise 
\end{cases}
$$</p>

another case was:

$$
f_s = \frac{1}{T_s}
$$

This too had a problem, thankful your underline option fixed it:

$$
f<em>s = \frac{1}T<em>s
$$

in short, its probably safer to turn off markup processing between "$" .. "$"  and "$$" .. "$$" tags....

however, you really should convert ampersands inside of these mathjax blocks... mathjax needs a raw ampersand... I would put that on your bug list... its quite common to use ampersands in mathjax script... you need  a way to pass a raw ampersand though your script when its in the math fence block.

**Comment by wm2015email** [comment]
Here's some code i was using to test your markdown engine in combination with mathjax:

```
<!DOCTYPE html>
<html>
  <head>
    <title>Read File (via User Input selection)</title>
	
	<!-- node.js packages required: -->
	<!--     npm install jquery -->
	<!--     npm install showdown -->
	<!--     npm install mathjax -->
	
    <script type="text/javascript" src="./node_modules/showdown/dist/jquery.js"></script>
	
    <script type="text/javascript" src="./node_modules/showdown/dist/showdown.js"></script>
	
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({
            tex2jax: {inlineMath: [["$","$"],["\\(","\\)"]]}
    });
    </script>
	
    <script type="text/javascript" 
	    src="./node_modules/mathjax/MathJax.js?config=TeX-AMS_HTML-full"></script>

	
    <script type="text/javascript">
    var reader; 

    function checkFileAPI() {
        if (window.File && window.FileReader && window.FileList && window.Blob) {
            reader = new FileReader();
            return true; 
        } else {
            alert('The File APIs are not fully supported by your browser. Fallback required.');
            return false;
        }
    }

    function readText(filePath) {
        var output = ""; //placeholder for text output
        if(filePath.files && filePath.files[0]) {           
            reader.onload = function (e) {
                output = e.target.result;
                displayContents(output);
            };//end onload()
            reader.readAsText(filePath.files[0]);
        }//end if html5 filelist support
        else if(ActiveXObject && filePath) { //fallback to IE 6-8 support via ActiveX
            try {
                reader = new ActiveXObject("Scripting.FileSystemObject");
                var file = reader.OpenTextFile(filePath, 1); //ActiveX File Object
                output = file.ReadAll(); //text contents of file
                file.Close(); //close file "input stream"
                displayContents(output);
            } catch (e) {
                if (e.number == -2146827859) {
                    alert('Unable to access local files due to browser security settings. ' + 
                     'To overcome this, go to Tools->Internet Options->Security->Custom Level. ' + 
                     'Find the setting for "Initialize and script ActiveX controls not marked as safe" and change it to "Enable" or "Prompt"'); 
                }
            }       
        }
        else { //this is where you could fallback to Java Applet, Flash or similar
            return false;
        }       
        return true;
    }   

    function displayContents(txt) {
        showdown.setOption('underline', true);
		converter = new showdown.Converter({underline: true});
        showdown.setOption('underline', true);
		html = converter.makeHtml(txt);
        var el = document.getElementById('main'); 
        el.innerHTML = html; //display output in DOM

        MathJax.Hub.Queue(["Typeset",MathJax.Hub]);
    }   
</script>
</head>

<body onload="checkFileAPI();">
    <div id="container">    
        <input type="file" onchange='readText(this)' />
        <br/>
        <hr/>   
		
        <h3>Contents of the Text file:</h3>
		
        <div id="main">
            ...
        </div>
    </div>
</body>

</html>
```

**Comment by wm2015email** [comment]
Here's a way to fix it that's relatively easy... that's easy to build "mathjax" option into your markdown library... just base64 encode anything inbetween $ .. $  and $$ .. $$...and then base64 decode when you are done...  its a little bit messy however, because i'm not 100% sure about the latex rules for preventing runaway uncloced math dollar quotes.  but i think its something like this:

```


console.log("start")


function btoa(s) {
	s = s.replace(/[^\x00-\x7F]/g, "");
    return Buffer.from(s).toString('base64');
}

function atob(s) {
	return Buffer.from(s, 'base64').toString();
}


function isEmptyOrSpaces(str){
    return str === null || str.match(/^[\s\n\r]*$/) !== null;
}

// this will either base64 encode or base64 decode what's between math dollar quotes.
function mathjax_coder(mode, s) {
    //console.log("mathjax_encode: mode:"  + mode);
    //console.log("length = "  + s.length);
    	
	linenumber = 1;
	
    if (!(mode == 'e' || mode == 'd')) {
	   throw "invalid mode, required: e=encode, d=decode";
	}
	
    var ostr  = "";
	var enc   = "";
	var state = 0;
	var line  = "";
	var dcount = 0; // math inlines per line
	var ddcount = 0;
    for(i=0; i < s.length; i++) {
	    c = s.charAt(i);
	    line += c;
			
		// blank line terminates open math dollar
		if (c == '\n' && isEmptyOrSpaces(line) && state != 0) {
	        console.log("WARNING: (" + mode + ") math dolar not terminated:" + linenumber-1 + ")" );
			state = 0;		    		
		}

	    // run away inline
	    if ((c == '\n') && (state != 0) && type == 1) {
	        console.log("WARNING: (" + mode + ") runaway inline math dollar not balanced at line:" + linenumber );
			state = 0;
		}
		
	    // $$ pair + another $ pair not allowed on the same line
	    if ((c == '\n') && (ddcount > 1)) {
	        console.log("WARNING: (" + mode + ") only one \$\$ pair allowed per line at  (line:" + linenumber + ")" );
			state = 0;		    
		}
		
		
	    if (c == '\n') {
		    linenumber++;
			line = "";
			dcount=0;
			ddcount=0;
		}
		
  	    //console.log(i + " " + c + "\n");
        switch(state) {
	   
	    // non-encoding
	    case 0:
		    // start marker1
            if (c == '$') {
	            state = 1;
				type  = 1;
		        enc   = "";
			    ostr +=  c; 
				dcount++;
		    }
			
			// non-encoded
			else {
	            state = 0;
   		        enc   = "";
			    ostr +=  c;
			}
	        break
		  
		// marker1
		case 1:
		    // start marker2
		    if (s.charAt(i) == '$') {
			    ostr +=  c; 
			    state = 3;
				type = 2;
				ddcount++;
			}
			// start encoding
			else {
			    type   = 1;
			    enc   += c;
			    state  = 3;
			}
			break;
	   
        //encoding
        case 3:
	        // stop marker1
		    if (s.charAt(i) == '$') {
				   if (type == 1) {
			  //       console.log("ENC1: " + enc)
					   state = 0;
			           ostr += (mode == 'e') ? btoa(enc) : atob(enc);
					   ostr += "$";
				   }
				   else {
					   state = 4;
				   }
			}
			else {
			    state  = 3;
			    enc   += c;
			}
		    
            break;		

	    // stop marker2
 	    case 4: 
	        if (s.charAt(i) != '$') {
			    console.log("WARNING: (" + mode + ") closing math dollar not balanced at line:" + linenumber );
				console.log("LINE:" + line);
				console.log("HERE:(" + enc + ")");
				//process.exit(1);
		    }
			state = 0;
	       //onsole.log("ENC2: " + enc)
	 		ostr += (mode == 'e') ? btoa(enc) : atob(enc);
			ostr += "$";
			ostr += "$";
            break;
			

		
        } //switch	
		
    } // for loop
	
	//console.log("OSTR:" + ostr)
	return ostr;

}


function MathJaxWrapper(html) {
    h = "\n";
    h += "<!DOCTYPE html>\n";
    h += "<html>\n";
    h += "<head>\n";
    h += "   <title>Markdown TexDollar Reader</title>\n";
    h += "   <script type='text/x-mathjax-config'>\n";
    h += "        MathJax.Hub.Config({\n";
    h += "          tex2jax: {inlineMath: [['$','$']]}\n";
    h += "    });\n";
    h += "    </script>\n";
    h += "    <script type='text/javascript'\n"; 
    h += "	    src='./node_modules/mathjax/MathJax.js?config=TeX-AMS_HTML-full'></script>\n";
    h += "</head>\n";
    h += "\n";
    h += "<body>\n";  
    h += html;
    h += "</body>\n";
    h += "</html>\n";   
	return h;
}


//t1 = btoa("this is a test");
//t2 = atob(t1);
//
//console.log(t1 + "\n");
//console.log(t2 + "\n");
//process.exit(1);

var filename  = "out2.html";

var fs        = require('fs');
var showdown  = require('showdown'); 

showdown.setOption('underline', true);

var file = 'test1.md';

var mark1  = fs.readFileSync(file, 'utf8');
var mark2  = mathjax_coder('e', mark1);

fs.writeFileSync("mark2.md", mark2);

var converter = new showdown.Converter();
var h1    = converter.makeHtml(mark2);
var h2    = mathjax_coder('d', h1);
var JaxHtml   = MathJaxWrapper(h2);

console.log("Write File: " + filename);
fs.writeFileSync(filename, JaxHtml);

console.log("exit!");
process.exit();
```

**Comment by obedm503** [comment]
if you don't specifically need mathjax, you could use katex with the [showdown-katex](https://github.com/obedm503/showdown-katex) extension and won't have to wait for a change in showdown itself. katex does basically everything mathjax does but faster

**Comment by wm2015email** [comment]
thanks.  that's a good tip.
